todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/domain/logic.kt

```diff
@@ -0,0 +1,140 @@
+/*
+ * -\-\-
+ * --
+ * Copyright (c) 2017-2018 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.example.android.architecture.blueprints.todoapp.tasks.domain
+
+import com.example.android.architecture.blueprints.todoapp.tasks.domain.FeedbackType.*
+import com.example.android.architecture.blueprints.todoapp.util.change
+import com.example.android.architecture.blueprints.todoapp.util.without
+import com.spotify.mobius.Effects.effects
+import com.spotify.mobius.First
+import com.spotify.mobius.First.first
+import com.spotify.mobius.Next
+import com.spotify.mobius.Next.*
+
+
+fun init(model: TasksListModel) : First<TasksListModel, TasksListEffect> =
+        model.tasks?.let { first(model, effects(LoadTasks as TasksListEffect)) }
+                ?: first(model.copy(loading = true), effects(RefreshTasks, LoadTasks))
+
+fun update(model: TasksListModel, event: TasksListEvent): Next<TasksListModel, TasksListEffect> =
+        when (event) {
+            RefreshRequested -> onRefreshRequested(model)
+            NewTaskClicked -> onNewTaskClicked()
+            is NavigateToTaskDetailsRequested -> onNavigateToTaskDetailsRequested(model, event)
+            is TaskMarkedComplete -> onTaskCompleted(model, event)
+            is TaskMarkedActive -> onTaskActivated(model, event)
+            ClearCompletedTasksRequested -> onCompletedTasksCleared(model)
+            is FilterSelected -> onFilterSelected(model, event)
+            is TasksLoaded -> onTasksLoaded(model, event)
+            TaskCreated -> onTaskCreated()
+            TasksRefreshed -> onTasksRefreshed(model)
+            TasksFreshFailed -> onTasksRefreshFailed(model)
+            TasksLoadingFailed -> onTasksLoadingFailed(model)
+        }
+
+private fun onRefreshRequested(model: TasksListModel): Next<TasksListModel, TasksListEffect> =
+        next(model.copy(loading = true), effects(RefreshTasks))
+
+private fun onNewTaskClicked(): Next<TasksListModel, TasksListEffect> =
+        dispatch(effects(StartTaskCreationFlow))
+
+private fun onNavigateToTaskDetailsRequested(model: TasksListModel,
+                                             event: NavigateToTaskDetailsRequested)
+        : Next<TasksListModel, TasksListEffect> =
```

Comment:
I think this would be more readable as:
```kotlin
val task = model.findTaskById(event.taskId) ?: throw IllegalStateException("Task does not exist")        

dispatch(effects(NavigateToTaskDetails(it)))
```
But also, `findTaskById` already throws an exception if I'm reading it correctly, so you could probably just return a non-optional task there